toolchain: alias restrict to __restrict for GNUC

nslowell opened the pull request:
restrict keyword is not supported in C++, however
GNUC does support __restrict so we can use that.

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

nslowell committed:
toolchain: define ZRESTRICT for restrict aliasing

restrict keyword is not supported in C++, however
GNUC does support __restrict for C/C++ so we can use that,
however the complexity of toolchains requires a ZRESTRICT
symbol that should be used throughout the codebase to not
cause any conflicts or redefinition errors

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

pfalcon commented:
I'm not sure it's a scalable solution to patch each and every prototype using "restrict" to be in 2 duplicated versions.

Instead, some common header should define something like ZRESTRICT, and define it to either supported version of "restrict" keyword, or to nothing.

nslowell commented:
Understood.  Any location suggestions? kernel.h? toolchain.h?

pfalcon commented:
> Understood. Any location suggestions? kernel.h? toolchain.h?

toolchain.h sounds like a good idea, as this feature (supporting restrict for C++) may be inherently toolchain-specific, and there're talks/work on supporting more toolchains beyond GCC/LLVM. The best idea would be to look up a few committers to that file and invite them to the discussion.

nslowell commented:
Actually found a much simpler solution, let me know what you think.

andyross reviewed (commented):
Pedantry: this patch has nothing to do with the POSIX layer or semaphores.  It's a toolchain compatibility fix.   Label it "toolchain: gcc:" or whatever.

nslowell commented:
Sorry. I changed the commit message, forgot about the open pull request title.

andyross reviewed (approved):
Oh sorry, I don't care about the PR title, I was clearly just reading the wrong spot.

pfalcon commented:
@nslowell: As you can see from the build failures, it's not as simple as that. I also would do it like that for a personal project of me, but for a project the scale of Zephyr, I'm afraid the only sustainable solution is to indeed define a Zephyr-private (namespaced) preprocessor symbol, and use that throughout the codebase.

>                  from /workdir/zephyr/lib/open-amp/resource_table.c:30:
modules/libmetal/libmetal/lib/include/metal/compiler/gcc/compiler.h:19: error: "restrict" redefined [-Werror]
   19 | #define restrict __restrict__

nslowell committed:
posix: semaphore: use ZRESTRICT

To better handle all possible definitions of restrict
or __restrict keyword and avoid redefinition errors, use
the ZRESTRICT pre-processor defined in toolchain.h

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

nslowell commented:
Thanks for the continued help.  Hopefully this solves the issue.  Note there are now two different commits.

nashif commented on include/toolchain.h:
```diff
@@ -35,13 +35,16 @@
+#define ZRESTRICT restrict
 #elif defined(__GNUC__) || (defined(_LINKER) && defined(__GCC_LINKER_CMD__))
 #include <toolchain/gcc.h>
+#define ZRESTRICT __restrict
```
this should be defined in the respective toolchain header file, not in this top-level header.

nslowell replied:
xcc.h actually includes gcc.h which includes common.h.  Also, other.h not provided, so I need some more clarification on the best way to do this.

Is it ok for #define ZRESTRICT to stay along side other.h in this top-level header?
Do I just #define ZRESTRICT __restrict in gcc.h and let that take care of gcc and xcc?
Do I then #ifndef #define ZRESTRICT restrict #endif in common.h?  Will that be enough?

maxbachmann replied:
@nashif whats still needed to get this merged? It would be great to have this fixed

nslowell commented:
Not sure what is needed to get this merged.

pfalcon commented:
@nashif: re: your comment https://github.com/zephyrproject-rtos/zephyr/pull/27245#pullrequestreview-467267334 :

> this should be defined in the respective toolchain header file, not in this top-level header.

This comment was addressed. Now, respective toolchain files define `ZRESTRICT`, though there's still a fallback in include/toolchain.h, where ZRESTRICT is defined to "natural default". So, if someone drops in a new toolchain-specific file without ZRESTRICT, they still will get a good-enough default. IMHO, that's reasonable approach. Can you please re-review?

Additional note about PR "assignees". The original idea behind that was that assigned person will "help to drive a patch to the completion/inclusion". But soon after it was launched, turned out, it works rather differently. How it works is that other merge-drivers (i.e. people with merge rights) start to bypass it. Certainly that's done from good motives of "not stomping on somebody else's feet". But the end result we see here (and I saw similar cases elsewhere (of course, not just with you)) is that you self-assigned this ticket on Aug 14, and since then, there was no further review activity happening on it. So IMHO, the process of PR assignees might use some adjustment.

Thanks.

pfalcon commented:
Retriggering CI.

pfalcon commented:
@nslowell: Actually, if you could rebase on the latest master, that would be helpful, as I don't really see Buildkite to running a new build (and if it does, dunno if it tries to apply a PR on the latest master, or uses whatever branch was there originally.)

nslowell commented:
@nashif please confirm change request fulfilled. @pfalcon this was updated as requested but no action was taken.

pfalcon commented:
@nslowell :

> @nashif please confirm change request fulfilled. @pfalcon this was updated as requested but no action was taken.

I myself +1'ed this PR. I don't have merge permission to merge it, and it can't be, until there's even a single -1 vote, as still presented from @nashif. So pinging him to lift it is the right thing. (And thanks for not dropping this PR, it's worth being merged IMHO.)

bbolen commented:
@nashif Can you take a look at this PR again and see if you can remove your -1?

stephanosio commented on include/toolchain/gcc.h:
```diff
@@ -63,6 +63,8 @@
 #define BUILD_ASSERT(EXPR, MSG...)
 #endif
 
+#define ZRESTRICT __restrict
```
Should use the C standard keyword when not compiling C++ code.
```suggestion
#ifdef __cplusplus
#define ZRESTRICT __restrict
#else
#define ZRESTRICT restrict
#endif
```

nslowell replied:
The only reason to switch would be if someone is disabling GNU extensions which would most likely break other things during compilation with gcc, otherwise there is no difference in the behavior, so a single line is much cleaner than yet another conditional macro.

stephanosio commented on include/toolchain/common.h:
```diff
@@ -26,6 +26,10 @@
 #define TASK_ENTRY_CPP  extern "C"
 #endif
 
+#ifndef ZRESTRICT
+#define ZRESTRICT restrict
+#endif
```
This is not correct when compiling C++ code (i.e. when `__cplusplus` is defined).

If toolchain-specific C++ restrict qualifier is not defined, the default behaviour should be to use `restrict` keyword for C and ignore restrict qualification for C++ as the C++ standard does not support such a qualifier.

```suggestion
#ifndef ZRESTRICT
#ifndef __cplusplus
#define ZRESTRICT restrict
#else
#define ZRESTRICT
#endif
#endif /* !defined(ZRESTRICT) */
```

stephanosio reviewed (changes requested):
Should update/get rid of `_MLIBC_RESTRICT` now that a global macro is available:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/libc/minimal/include/bits/restrict.h#L19-L23

Also this is just a suggestion, but I wonder if we should go with `RESTRICT` instead of `ZRESTRICT`. I do not recall seeing `RESTRICT` in any other projects ...

nslowell commented:
> Should update/get rid of `_MLIBC_RESTRICT` now that a global macro is available:
> https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/libc/minimal/include/bits/restrict.h#L19-L23

@stephanosio Can you give more explanation about what you mean? What global macro? _MLIBC_RESTRICT is used in a lot of places so it be safer to leave it alone.

nslowell commented on include/toolchain.h:
```diff
@@ -46,6 +46,9 @@
  * and therefore this header is not meant to exist in-tree
  */
 #include <toolchain/other.h>
+#ifndef ZRESTRICT
+#define ZRESTRICT restrict
+#endif
```
That is potentially a new requirement on old headers and more of a burden on those headers when it can be done for them by it being here.

Commit:
libc: minimal: Use new ZRESTRICT macro

This lets the toolchain header files determine how to use "restrict"
instead of having that decision down in the minimal libc library.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
The "Coding Guidelines" test failure is a false positive.  All other comments have been addressed.

stephanosio reviewed (approved):
Looks reasonable enough.
